src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs

```diff
@@ -35,13 +37,96 @@ public ValueTask<int> ReadAsync(Memory<byte> memory)
     }
 
     public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+    {
+        if (SegmentsOverChunksLimit(data))
+        {
+            return WriteDataOverChunksLimit(data);
+        }
+
+        return WriteDataAsync(data);
+    }
+
+    private ValueTask<int> WriteDataAsync(in ReadOnlySequence<byte> data)
     {
         var write = GetWriteOperation();
         write.Initialize(_handler, data);
         Run(write);
         return new ValueTask<int>(write, 0);
     }
 
+    /*
+    * In case the number of chunks is bigger than responseMaxChunks we need to make multiple calls 
+    * to the native api https://docs.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/ihttpresponse-writeentitychunks-method  
+    * Despite the documentation states that feeding the function with more than 65535 chunks will cause the function to throw an exception, 
+    * it actually seems that 65534 is the maximum number of chunks allowed.
+    * Also, there seems to be a problem when slicing a ReadOnlySequence on segment borders tracked here https://github.com/dotnet/runtime/issues/67607
+    * That's why we only allow 65533 chunks.  
```

Comment:
I hope I understood your question correctly but yeah, we do account for it precisely because of [that bug](https://github.com/dotnet/runtime/issues/67607) that I discovered while slicing the ROS. I added the comment so it's clear why the constant is not 65535 and I added it there because it's the part that do the slicing. Perhaps you are suggesting it's not the right place to have it?

Let me know if I didn't understand your question properly.

Reply:
This comment makes sense, I left this comment before I looked through how we are slicing this. We could conceivably avoid that bug by not slicing at all.

This should be possible by calling [ROS.TryGet](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1.tryget?view=net-6.0) instead of foreaching over the sequence and then using [this ROS ctor](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1.-ctor?view=net-6.0#system-buffers-readonlysequence-1-ctor(system-buffers-readonlysequencesegment((-0))-system-int32-system-buffers-readonlysequencesegment((-0))-system-int32)) to manually create the slice.

That way we'd definitely always have the correct segments count, and it would be way more efficient then slicing since the ROS struct is mutable. At first, I didn't want to suggest this since it'd be slightly more complicated than the foreach loop and we're not trying to really optimize `WriteDataOverChunksLimit`, but now I don't think this would that much more complicated. You don't have to do this, but it'd be nice.

Reply:
Very interesting, didn't know this, I am definitely interested in trying implementing it (with holidays in the middle it could take some time but I don't think too much)

Reply:
We don't have too much more time to get this in for rc1, and this code ends up being more complicated than I remembered because `TryGet` doesn't give you a `ReadOnlySequenceSegment<byte>`. You have to get it by taking the `ROS.Start` `SequencePosition` and assuming that is a `ReadOnlySequenceSegment<byte>` and then you can manually walk the linked list.

`Slice` already does basically that, and it should be a safe assumption that this cast is valid for any ROS backed by a default System.IO.Pipeline `PipeReader` implementation like we know this is, but it's no longer as simple.

https://github.com/dotnet/runtime/blob/865802a793a673625e4da4b3f7feb987a647d2c7/src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs#L230

So, I say just stick to the simple slice implementation after all.